ouroboros-consensus/src/Ouroboros/Consensus/Mempool/Update.hs

```diff
@@ -36,13 +37,86 @@ import qualified Ouroboros.Consensus.Mempool.TxSeq as TxSeq
   Add transactions
 -------------------------------------------------------------------------------}
 
+-- | Add a single transaction to the mempool, blocking if there is no space.
+--
+implAddTx ::
+     ( MonadSTM m
+     , MonadMVar m
+     , LedgerSupportsMempool blk
+     , HasTxId (GenTx blk)
+     )
+  => StrictTVar m (InternalState blk)
+     -- ^ The InternalState TVar.
+  -> MVar m () -- ^ The FIFO for remote peers
+  -> MVar m () -- ^ The FIFO for all remote peers and local clients
+  -> LedgerConfig blk
+     -- ^ The configuration of the ledger.
+  -> (GenTx blk -> TxSizeInBytes)
+     -- ^ The function to calculate the size of a
+     -- transaction.
+  -> Tracer m (TraceEventMempool blk)
+     -- ^ The tracer.
+  -> AddTxOnBehalfOf
+  -> GenTx blk
+     -- ^ The transaction to add to the mempool.
+  -> m (MempoolAddTxResult blk)
+implAddTx istate remoteFifo allFifo cfg txSize trcr onbehalf tx =
+    -- To ensure fair behaviour between threads that are trying to add
+    -- transactions, we make them all queue in a fifo. Only the one at the head
+    -- of the queue gets to actually wait for space to get freed up in the
+    -- mempool. This avoids small transactions repeatedly squeezing in ahead of
+    -- larger transactions.
+    --
+    -- The fifo behaviour is implemented using a simple MVar. And take this
+    -- MVar lock on a transaction by transaction basis. So if several threads
+    -- are each trying to add several transactions, then they'll interleave at
+    -- transaction granularity, not batches of transactions.
+    --
+    -- To add back in a bit of deliberate unfairness, we want to prioritise
+    -- transactions being added on behalf of local clients, over ones being
+    -- added on behalf of remote peers. We do this by using a pair of mvar
+    -- fifos: remote peers must wait on both mvars, while local clients only
+    -- need to wait on the second.
+    case onbehalf of
+      AddTxForRemotePeer ->
```

Comment:
We are waiting forever here, would it not make sense to introduce some timeout? Same for the local tx submission.

Reply:
One option is that this timeout is implemented outside the mempool API. The other question is in which circumstances clients will wait forever (ie, how realistic is this?).

Reply:
_In theory, there is no difference between theory and practice. In practice, there is_

Reply:
I don't think clients will wait *forever*, but it seems to me that having a blocking call in a queue is not great from a UX perspective: If I submit a tx locally, I would be interested in having an upper bound on the time it can take for my tx to get in, and let me retry if this timeout expires.

Reply:
I'm not sure I agree @abailly-iohk . I'd much rather the node handle the retry for me until my TTL expires and that tx is no longer valid on-chain. For example, shortly after alonzo, there was a decent amount of congestion. At the time, I had to send a few thousand transactions. I found it much nicer to just let it go for 24 hours until all the tx got adopted.

Reply:
Fair enough, I guess different use cases entail different strategies. My point is that right now there's no retry in the proposed fix, the client is  just _blocked_ and it could be for hours depending on how transactions propagate across the network. I have not done a deep analysis of how this new behaviour will impact propagation of txs and blocks across the network so perhaps there's a counter-argument to be made that if all clients are blocked once the network is congested, this might apply sufficient backpressure to reduce congestion more swiftly. OTOH the mempools might be filled with many different txs which might take more or less time to end up in a block should _all_ mempools be filled.
At this point, I would be interested in having some kind of more formal analysis on how the network would behave. I could work on this myself but I guess other people might have a deeper expertise to produce more significant result more quickly than I could do.

Reply:
There would be no benefit in that circumstance from failing and retrying (and indeed you'd go to the back of the queue).

Local clients can already abandon a tx submission if they want to, just by terminating the protocol. But as I say, it's not a good idea. It doesn't provide any more feedback or any better progress.

Reply:
Also, blocking when congested entails no extra work. Polling/retrying is busy work.

Once the whole network is congested, tx submission works like an event driven system: each time space is freed up then we make progress and send "events" (really requests for more work) out to our peers who are trying to submit txs.

Having pending work available immediately means this is quite efficient and there's no waiting.